node/auth/authenticator.go

```diff
@@ -0,0 +1,172 @@
+package auth
+
+import (
+	"context"
+	"fmt"
+	grpc "github.com/Layr-Labs/eigenda/api/grpc/node/v2"
+	"github.com/Layr-Labs/eigenda/core"
+	gethcommon "github.com/ethereum/go-ethereum/common"
+	lru "github.com/hashicorp/golang-lru/v2"
+	"time"
+)
+
+// RequestAuthenticator authenticates requests to the DA node. This object is thread safe.
+//
+// This class has largely been future-proofed for decentralized dispersers, with the exception of the
+// preloadCache method, which will need to be updated to handle decentralized dispersers.
+type RequestAuthenticator interface {
+	// AuthenticateStoreChunksRequest authenticates a StoreChunksRequest, returning an error if the request is invalid.
+	// The origin is the address of the peer that sent the request. This may be used to cache auth results
+	// in order to save node resources.
+	AuthenticateStoreChunksRequest(
+		ctx context.Context,
+		origin string,
+		request *grpc.StoreChunksRequest,
+		now time.Time) error
+}
+
+// keyWithTimeout is a key with that key's expiration time. After a key "expires", it should be reloaded
+// from the chain state in case the key has been changed.
+type keyWithTimeout struct {
+	key        gethcommon.Address
+	expiration time.Time
+}
+
+var _ RequestAuthenticator = &requestAuthenticator{}
+
+type requestAuthenticator struct {
+	// chainReader is used to read the chain state.
+	chainReader core.Reader
+
+	// keyCache is used to cache the public keys of dispersers.
```

Comment:
what public key has only 32 bits?

Reply:
The `uint32` is the relay ID. These are serial numbers, not eth keys. Currently we have exactly one disperser, and its ID is hard coded to be `uint32(0)`.